File: gui/iode_objs/edit/edit_table_model.cpp
#pragma once
#include "edit_table_model.h"

EditTableModel::EditTableModel(const QString& tableName, QWidget* parent) : QAbstractTableModel(parent)
{
	try
	{
		table = Tables.get(tableName.toStdString());
	}
	catch (const std::exception& e)
	{
		table = nullptr;
		QMessageBox::critical(static_cast<QWidget*>(parent), tr("ERROR"), tr(e.what()));
	}
}

EditTableModel::~EditTableModel()
{
	if(table) 
		delete table;
}

int EditTableModel::rowCount(const QModelIndex& parent) const
{
	// + 1 for the divider line
	return table->nb_lines() + 1;
}

int EditTableModel::columnCount(const QModelIndex& parent) const
{
	return table->nb_columns();
}

Qt::ItemFlags EditTableModel::flags(const QModelIndex& index) const
{
	if (!index.isValid())
		return Qt::ItemIsEnabled;

	return QAbstractItemModel::flags(index) | Qt::ItemIsEditable;
}

QVariant EditTableModel::headerData(int section, Qt::Orientation orientation, int role) const
{
	QString h_data;

	if (role != Qt::DisplayRole)
		return QVariant();

	if (orientation == Qt::Vertical)
	{
		if (section == 0)
		{
			h_data = "DIVIS";
		}
		else
		{
			TableLineType line_type = table->get_line(section - 1)->get_line_type();
			switch (line_type)
			{
			case TABLE_LINE_TITLE:
				h_data = "TITLE";
				break;
			case TABLE_LINE_CELL:
				h_data = "CELLS";
				break;
			case TABLE_LINE_MODE:
				h_data = "MODE";
				break;
			case TABLE_LINE_FILES:
				h_data = "FILES";
				break;
			case TABLE_LINE_DATE:
				h_data = "DATE";
				break;
			default:
				h_data = " ";
				break;
			}
		}
	}
	else
	{
		if (section == 0) h_data = "Line Titles";
		if (section == 1) h_data = "Formulas";
	}
	return QVariant(h_data);
}

QVariant EditTableModel::data(const QModelIndex& index, int role) const
{
	QString s_data;

	if (!index.isValid())
		return QVariant();

	bool is_divider = index.row() == 0;
	int nb_columns = table->nb_columns();
	TableLine* line = is_divider ? table->get_divider_line() : table->get_line(index.row() - 1);
	TableLineType line_type = line->get_line_type();
	TableCell* cell = (line_type == TABLE_LINE_TITLE || line_type == TABLE_LINE_CELL) ? line->get_cell(index.column(), nb_columns) : nullptr;

	if (role == Qt::TextAlignmentRole)
	{
		TableCellAlign cell_align = (cell != nullptr) ? cell->get_align() : TableCellAlign::TABLE_CELL_LEFT;
		switch (cell_align)
		{
		case TABLE_CELL_LEFT:
			return int(Qt::AlignLeft);
		case TABLE_CELL_CENTER:
			return int(Qt::AlignCenter);
		case TABLE_CELL_RIGHT:
			return int(Qt::AlignRight);
		case TABLE_CELL_DECIMAL:
			return int(Qt::AlignRight);
		default:
			return int(Qt::AlignLeft);
		}
	}

	if (role == Qt::FontRole)
	{
		if (is_divider) 
			return int(QFont::StyleNormal);
		
		// line of type FILES, MODE, LINE or DATE
		if (!cell) 
			return int(QFont::StyleNormal);
		
		QFont font;
		if (cell->is_bold()) 
			font.setBold(true);
		if (cell->is_italic()) 
			font.setItalic(true);
		if (cell->is_underline()) 
			font.setUnderline(true);
		return font;
	}

	if (role == Qt::BackgroundRole)
	{
		QColor color;
		switch (line_type)
		{
		case TABLE_LINE_SEP:
			color = QColor(Qt::black);
			break;
		case TABLE_LINE_CELL:
			color = QColor(Qt::white);
			break;
		default:
			color = QColor(Qt::lightGray);
			color.setAlphaF(0.4);
			break;
		}
		return QVariant(color);
	}

	if (role == Qt::DisplayRole || role == Qt::EditRole)
	{
		QString s_data;
		bool display_quotes; 
		switch (line_type)
		{
		case TABLE_LINE_TITLE:
			s_data = QString::fromStdString(cell->get_content(false));
			break;
		case TABLE_LINE_CELL:
			display_quotes = (!is_divider) && cell->get_type() == TABLE_CELL_STRING;
			s_data = QString::fromStdString(cell->get_content(display_quotes));
			break;
		default:
			break;
		}
		emit displayData(index.row(), line_type, columnCount());
		return QVariant(s_data);
	}

	return QVariant();
}

bool EditTableModel::setData(const QModelIndex& index, const QVariant& value, int role)
{
	if (index.isValid() && role == Qt::EditRole)
	{
		std::string std_value = value.toString().toStdString();
		try
		{
			bool is_divider = index.row() == 0;
			int nb_columns = table->nb_columns();
			TableLine* line = is_divider ? table->get_divider_line() : table->get_line(index.row() - 1);
			TableLineType line_type = line->get_line_type();
			TableCell* cell = (line_type == TABLE_LINE_TITLE || line_type == TABLE_LINE_CELL) ? line->get_cell(index.column(), nb_columns) : nullptr;

			if (is_divider)
			{
				if (cell->get_type() == TABLE_CELL_STRING)
					cell->set_text(std_value);
				else
					cell->set_lec(std_value);
			}
			else
			{
				TableLineType line_type = line->get_line_type();
				switch (line_type)
				{
				case TABLE_LINE_TITLE:
					cell->set_text(std_value);
					break;
				case TABLE_LINE_CELL:
					// When inserting a new line of type TABLE_LINE_CELL, the attribute TCELL::tc_type of cells is undefined!
					// Rule: If the content starts with a double quotes -> we assume it is a string cell. 
					//       Otherwise, it is a LEC cell.
					cell->set_content(std_value);
					break;
				default:
					break;
				}
			}
			emit dataChanged(index, index, { role });
			return true;
		}
		catch (const std::exception& e)
		{
			QMessageBox::warning(nullptr, "ERROR", QString(e.what()));
		}
	}
	else
	{
		return false;
	}
}

void EditTableModel::appendLine(TableLineType lineType)
{
	switch(lineType)
	{
	case TableLineType::TABLE_LINE_CELL:
		// When adding a new line of type TABLE_LINE_CELL, the attribute TCELL::tc_type of cells is undefined!
		table->add_line_with_cells();
		break;
	case TableLineType::TABLE_LINE_TITLE:
		table->add_title("");
		break;
	case TableLineType::TABLE_LINE_SEP:
		table->add_line_separator();
		break;
	case TableLineType::TABLE_LINE_FILES:
		table->add_line_files();
		break;
	case TableLineType::TABLE_LINE_DATE:
		table->add_line_date();
		break;
	case TableLineType::TABLE_LINE_MODE:
		table->add_line_mode();
		break;
	default:
		break;
	}

	resetModel();
}

void EditTableModel::insert_line(TableLineType lineType, const int position, const bool after)
{
	switch(lineType)
	{
	case TableLineType::TABLE_LINE_CELL:
		// When inserting a new line of type TABLE_LINE_CELL, the attribute TCELL::tc_type of cells is undefined!
		table->insert_line_with_cells(position, after);
		break;
	case TableLineType::TABLE_LINE_TITLE:
		table->insert_title(position, "", after);
		break;
	case TableLineType::TABLE_LINE_SEP:
		table->insert_line_separator(position, after);
		break;
	case TableLineType::TABLE_LINE_FILES:
		table->insert_line_files(position, after);
		break;
	case TableLineType::TABLE_LINE_DATE:
		table->insert_line_date(position, after);
		break;
	case TableLineType::TABLE_LINE_MODE:
		table->insert_line_mode(position, after);
		break;
	default:
		break;
	}

	resetModel();
}

bool EditTableModel::removeRows(int position, int rows, const QModelIndex& index)
{
	beginRemoveRows(QModelIndex(), position, position + rows - 1);

	try
	{
		// divider line doesn't count -> real table row = gui table row - 1
		for(int row = position; row < position + rows; row++)
			table->delete_line(row - 1);
	}
	catch (const std::exception& e)
	{
		QMessageBox::warning(nullptr, "WARNING", QString(e.what()));
		return false;
	}

	endRemoveRows();
	return true;
}

TableLineType EditTableModel::get_line_type(const int row) const
{
	// first row represents the divider line
	TableLine* line = (row == 0) ? table->get_divider_line() : table->get_line(row - 1);
	return line->get_line_type();
}

File: gui/iode_gui/iode_objs/edit/edit_table_model.py
from PySide6.QtCore import QModelIndex, Qt, Signal, QAbstractTableModel
from PySide6.QtWidgets import QMessageBox, QWidget
from PySide6.QtGui import QFont, QColor

from iode import Table, TableLineType, TableCellAlign, tables
from iode.util import TableLine, TableCell

# NOTE FOR THE DEVELOPERS:
# Read-only model: https://doc.qt.io/qt-6/model-view-programming.html#read-only-access
# Editable model:  https://doc.qt.io/qt-6/model-view-programming.html#editable-items
# Resizable model: https://doc.qt.io/qt-6/model-view-programming.html#resizable-models


class EditTableModel(QAbstractTableModel):

    display_data = Signal(int, int, int)

    def __init__(self, table_name: str, parent=None):
        super().__init__(parent)
        try:
            self._table: Table = tables[table_name]
        except Exception as e:
            self._table = None
            QMessageBox.critical(parent, "ERROR", str(e))

    @property
    def table(self) -> Table:
        return self._table

    def line_type(self, row: int) -> TableLineType:
        return TableLineType[self._table[row - 1].line_type]

    def rowCount(self, parent: QModelIndex=QModelIndex()) -> int:
        return len(self._table) + 1

    def columnCount(self, parent: QModelIndex=QModelIndex()) -> int:
        return self._table.nb_columns

    def flags(self, index: QModelIndex) -> Qt.ItemFlag:
        if not index.isValid():
            return Qt.ItemFlag.ItemIsEnabled
        return super().flags(index) | Qt.ItemFlag.ItemIsEditable

    def headerData(self, section: int, orientation: Qt.Orientation, role: Qt.ItemDataRole):
        if role != Qt.ItemDataRole.DisplayRole:
            return None

        if orientation == Qt.Orientation.Vertical:
            if section == 0:
                return "DIVIS"
            else:
                line_type: TableLineType = TableLineType[self._table[section - 1].line_type]
                return " " if line_type == TableLineType.SEP else line_type.name
        else:
            if section == 0:
                return "Line Titles"
            elif section == 1:
                return "Formulas"
            else:
                return " "

    def data(self, index: QModelIndex, role: Qt.ItemDataRole=Qt.ItemDataRole.DisplayRole):
        if not index.isValid():
            return None

        is_divider = index.row() == 0
        line: TableLine = self._table.divider if is_divider else self._table[index.row() - 1]
        line_type: TableLineType = TableLineType[line.line_type]
        cell: TableCell = None
        if line_type == TableLineType.TITLE:
            cell = line[0]
        elif line_type == TableLineType.CELL:
            cell = line[index.column()]

        if role == Qt.ItemDataRole.TextAlignmentRole:
            cell_align: TableCellAlign = TableCellAlign[cell.align] if cell else TableCellAlign.LEFT 
            if cell_align == TableCellAlign.LEFT:
                return int(Qt.AlignmentFlag.AlignLeft)
            elif cell_align == TableCellAlign.CENTER:
                return int(Qt.AlignmentFlag.AlignCenter)
            elif cell_align == TableCellAlign.RIGHT or cell_align == TableCellAlign.DECIMAL:
                return int(Qt.AlignmentFlag.AlignRight)
            else:
                return int(Qt.AlignmentFlag.AlignLeft)

        if role == Qt.ItemDataRole.FontRole:
            font = QFont()
            if is_divider or not cell:
                return font
            else:
                if cell.bold:
                    font.setBold(True)
                if cell.italic:
                    font.setItalic(True)
                if cell.underline:
                    font.setUnderline(True)
                return font

        if role == Qt.ItemDataRole.BackgroundRole:
            if line_type == TableLineType.SEP:
                color = QColor("black")
            elif line_type == TableLineType.CELL:
                color = QColor("white")
            else:
                color = QColor("lightGray")
                color.setAlphaF(0.4)
            return color

        if role in (Qt.ItemDataRole.DisplayRole, Qt.ItemDataRole.EditRole):
            if line_type == TableLineType.TITLE or line_type == TableLineType.CELL:
                str_data = str(cell)
            else:
                str_data = " "
            self.display_data.emit(index.row(), line_type, self.columnCount())
            return str_data

        return None

    def setData(self, index: QModelIndex, value, role: Qt.ItemDataRole):
        if index.isValid() and role == Qt.ItemDataRole.EditRole:
            value = str(value).strip()

            try:
                is_divider = index.row() == 0
                line: TableLine = self._table.divider if is_divider else self._table[index.row() - 1]
                line_type: TableLineType = TableLineType[line.line_type]

                if line_type == TableLineType.TITLE: 
                    if not value.startswith('"'): 
                        value = '"' + value
                    if not value.endswith('"'):
                        value = value + '"'
                    line[0] = value
                elif line_type == TableLineType.CELL:
                    line[index.column()] = value
                else:
                    pass

                self.dataChanged.emit(index, index, [role])
                return True
            except Exception as e:
                QMessageBox.warning(None, "ERROR", str(e))

        return False

    def append_line(self, line_type: TableLineType):
        if line_type == TableLineType.TITLE:
            self._table += " "
        elif line_type == TableLineType.CELL:
            self._table += [''] * self._table.nb_columns
        else:
            self._table += line_type
        self.reset_model()

    def insert_line(self, line_type: TableLineType, position: int, after: bool):
        if after:
            position += 1
        if line_type == TableLineType.TITLE:
            self._table.insert(position, " ")
        elif line_type == TableLineType.CELL:
            self._table.insert(position, [''] * self._table.nb_columns)
        else:
            self._table.insert(position, line_type)
        self.reset_model()

    def removeRows(self, position: int, rows: int, index: QModelIndex=QModelIndex()):
        self.beginRemoveRows(QModelIndex(), position, position + rows - 1)
        try:
            # divider line doesn't count -> real table row = gui table row - 1
            for row in range(position, position + rows):
                del self._table[row - 1]
            self.endRemoveRows()
        except Exception as e:
            QMessageBox.warning(None, "WARNING", str(e))
            self.endRemoveRows()
            return False
        return True

    def reset_model(self):
        self.beginResetModel()
        self.endResetModel()
